### PR TITLE
refactor(impala): port to sqlglot

### PR DESCRIPTION
**Pull request by cpcloud**
## Description of changes

Ports the impala backend to use the sqlglot compiler.

The main API change here is that the `add_operation` and `register` APIs for
Impala UDFs are removed.

Instead, you can simply declare the function using the existing APIs
`scalar_function`/`aggregate_function` and immediately call them.

These use the newer `udf.scalar.builtin` and `udf.agg.builtin` functions
internally, which reduces UDF machinery further.

I haven't yet removed the code in `ibis/backends/base/sql/registry` since that is still
in use by the Flink backend (and BigQuery until #7968 is merged). Once we port the Flink backend
over we can delete that code.

This PR may also serve as good starting point for a custom sqlglot dialect.

BREAKING CHANGE: UDFs no longer require explicit registration. Remove any calls to `Function.register`. If you were passing `database` to `Function.register`, pass that to `scalar_function` or `aggregate_function` as appropriate.

### CODE REVIEW

File: ibis/expr/types/joins.py [hunk 1]
```diff
@@ -81,15 +81,15 @@ def dereference_sides(left, right, deref_left, deref_right):
...
+def dereference_comparison_op(pred, deref_left, deref_right):
     left, right = dereference_sides(pred.left, pred.right, deref_left, deref_right)
     return pred.copy(left=left, right=right)
 
 
 def dereference_value(pred, deref_left, deref_right):
     deref_both = {**deref_left, **deref_right}
-    if isinstance(pred, ops.Binary) and pred.left.relations == pred.right.relations:
-        return dereference_binop(pred, deref_left, deref_right)
+    if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
```
Comment (kszucs) [inline ibis/expr/types/joins.py:91]: Can you pull this change out with a failing test case?

Comment (cpcloud) [reply ibis/expr/types/joins.py:91]: The test case included here is failing without my change.

File: ibis/backends/base/sql/__init__.py [hunk 2]
```diff
@@ -22,7 +22,6 @@
     import pandas as pd
     import pyarrow as pa
 
-raise RuntimeError("Temporarily make the SQL backends dysfunctional")
```
Comment (kszucs) [inline ibis/backends/base/sql/__init__.py:25]: Can we keep it while all the backends get ported?

Comment (cpcloud) [reply ibis/backends/base/sql/__init__.py:25]: Not really, I need it to be able to use udfs in impala

File: ibis/backends/base/sql/ddl.py
Comment (kszucs) [inline ibis/backends/base/sql/ddl.py:1]: Do we use this file somewhere?

File: ibis/expr/types/joins.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/types/joins.py:91]: Done in https://github.com/ibis-project/ibis/pull/7996/

File: ibis/backends/base/sqlglot/compiler.py [hunk 3]
```diff
@@ -822,6 +822,14 @@ def visit_Window(self, op, *, how, func, start, end, group_by, order_by):
         end_value = end.get("value", "UNBOUNDED")
         end_side = end.get("side", "FOLLOWING")
 
+        if f"{start_value} {start_side}".lower() in ("0 following", "0 preceding"):
```
Comment (kszucs) [inline ibis/backends/base/sqlglot/compiler.py:825]: Could `if start_value == 0` be sufficient?

File: ibis/backends/datafusion/compiler.py [hunk 4]
```diff
@@ -102,6 +103,8 @@ def visit_NonNullLiteral(self, op, *, value, dtype):
             return self.f.date_trunc("day", value.isoformat())
         elif dtype.is_binary():
             return sg.exp.HexString(this=value.hex())
+        elif dtype.is_uuid():
+            return sge.convert(str(value))
```
Comment (kszucs) [inline ibis/backends/datafusion/compiler.py:107]: This is the same in the pyspark backend. Could this be moved to the base compiler perhaps?

File: ibis/backends/base/sql/ddl.py
Comment (cpcloud) [reply ibis/backends/base/sql/ddl.py:1]: Yes, a bunch of places in the impala backend.

File: ibis/backends/datafusion/compiler.py (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/datafusion/compiler.py:107]: A bunch of the backends have this implementation, but I'd rather not do this refactor now (or maybe ever). I don't think we lose much by repeating this particular 2 lines.

File: ibis/backends/impala/__init__.py [hunk 5]
```diff
@@ -259,16 +256,21 @@ def raw_sql(self, query: str):
...
+        return PandasData.convert_table(results, schema)
 
     @contextlib.contextmanager
     def _safe_raw_sql(self, query: str | DDL | DML):
         if not isinstance(query, str):
-            query = query.compile()
+            try:
+                query = query.sql(dialect=self.compiler.dialect)
+            except AttributeError:
+                query = query.compile()
```
Comment (kszucs) [inline ibis/backends/impala/__init__.py:271]: Is this for the `base/sql/ddl.py` objects?

Comment (cpcloud) [reply ibis/backends/impala/__init__.py:271]: Yes.

File: ibis/backends/impala/client.py
Comment (kszucs) [inline ibis/backends/impala/client.py:1]: I removed the `PySparkTable` in the pyspark PR. I think we should rethink the API for backend specific table types in the near future.

Comment (cpcloud) [reply ibis/backends/impala/client.py:1]: Agreed. This one has a lot of existing methods, so I'm not removing it in this PR.

File: ibis/backends/impala/tests/snapshots/test_exprs/test_filter_with_analytic/out.sql [hunk 6]
```diff
@@ -1,18 +1,16 @@
...
+  SELECT
+    `t1`.`col`,
+    COUNT(*) OVER (ORDER BY NULL ASC NULLS LAST) AS `analytic`
+  FROM (
+    SELECT
+      `t0`.`col`,
+      NULL AS `filter`
+    FROM `x` AS `t0`
+    WHERE
+      NULL IS NULL
```
Comment (kszucs) [inline ibis/backends/impala/tests/snapshots/test_exprs/test_filter_with_analytic/out.sql:14]: Simplification opportunity here. `filter` is nicely inlined, we could eliminate the whole filter. Not important here of course.

File: ibis/backends/impala/tests/snapshots/test_exprs/test_named_from_filter_group_by/foo.sql [hunk 7]
```diff
@@ -1,4 +1,17 @@
-SELECT t0.`key`, sum(((t0.`value` + 1) + 2) + 3) AS `foo`
-FROM `t0` t0
-WHERE t0.`value` = 42
-GROUP BY 1
\ No newline at end of file
+SELECT
+  `t1`.`key`,
+  SUM((
```
Comment (kszucs) [inline ibis/backends/impala/tests/snapshots/test_exprs/test_named_from_filter_group_by/foo.sql:3]: Can we change the newline strategy in sqlglot?

Comment (kszucs) [reply ibis/backends/impala/tests/snapshots/test_exprs/test_named_from_filter_group_by/foo.sql:3]: Adds quite some vertical space in the snapshots.

Comment (cpcloud) [reply ibis/backends/impala/tests/snapshots/test_exprs/test_named_from_filter_group_by/foo.sql:3]: Not that i'm aware of.

File: ibis/backends/tests/test_generic.py [hunk 8]
```diff
@@ -1441,7 +1438,18 @@ def test_try_cast_func(con, from_val, to_type, func):
...
-        param(slice(3, 0), lambda _: 0, id="[3:0]"),
+        param(
+            slice(3, 0),
+            lambda _: 0,
+            id="[3:0]",
+            marks=[
+                pytest.mark.never(
+                    ["impala"],
+                    raises=ImpalaHiveServer2Error,
+                    reason="impala doesn't support OFFSET without ORDER BY",
```
Comment (kszucs) [inline ibis/backends/tests/test_generic.py:1449]: Could placing an order by null work here?

File: ibis/backends/impala/compiler.py [hunk 9]
```diff
@@ -1,66 +1,421 @@
...
+            return None
+        return spec
+
+    @singledispatchmethod
+    def visit_node(self, op, **kw):
+        return super().visit_node(op, **kw)
+
+    @visit_node.register(ops.Literal)
+    def visit_Literal(self, op, *, value, dtype):
+        if value is None and dtype.is_binary():
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/impala/compiler.py:84]: Yep, `BINARY` isn't accepted by Impala in that position for whatever reason.

File: ibis/backends/impala/compiler.py [hunk 10]
```diff
@@ -1,66 +1,421 @@
...
+            # implementation wraps on over- and underflow
+            return sge.convert(value.isoformat())
+        elif dtype.is_string():
+            value = (
+                value
+                # Escape \ first so we don't double escape other characters.
+                .replace("\\", "\\\\")
+                # ASCII escape sequences that are recognized in Python:
+                # https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals
+                .replace("\a", "\\a")  # Bell
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/impala/compiler.py:179]: Probably, but only if that's how you spell them in those backends (it likely is)

File: ibis/backends/tests/test_generic.py (hunk 8 above)
Comment (cpcloud) [reply ibis/backends/tests/test_generic.py:1449]: Hm, perhaps, but I didn't try it.

**Review by kszucs — APPROVED** [review]
Nice refactor!

Just nits, otherwise it looks good to me!

File: ibis/backends/base/sqlglot/compiler.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/base/sqlglot/compiler.py:825]: I was able to simplify it to `getattr(start_value, "this", None) == "0"`. `start_value` is a sqlglot object, so I have to handle that.

File: ibis/backends/tests/test_generic.py (hunk 8 above)
Comment (cpcloud) [reply ibis/backends/tests/test_generic.py:1449]: I leave this for later.

File: ibis/backends/impala/compiler.py [hunk 11]
```diff
@@ -1,66 +1,421 @@
...
+
+    def _aggregate(self, funcname: str, *args, where):
+        if where is not None:
+            args = tuple(self.if_(where, arg, NULL) for arg in args)
+
+        return self.f[funcname](*args, dialect=self.dialect)
+
+    @staticmethod
+    def _minimize_spec(start, end, spec):
+        if start is None and (
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/impala/compiler.py]: I added an extensive comment. I think we have some follow up work to do around clarifying the difference between `rows=(None, None)` and `rows=(None, 0)`.

**Pull request merged**